content/en/docs/setup/production-environment/tools/kubeadm/create-cluster-kubeadm.md

```diff
@@ -84,6 +86,57 @@ kubeadm to tell it what to do. This crashloop is expected and normal.
 After you initialize your control-plane, the kubelet runs normally.
 {{< /note >}}
 
+#### Network setup
+
+kubeadm similarly to other Kubernetes components tries to find a usable IP on
+the network interface associated with the default gateway on a host. Such
+an IP is then used for the advertising and/or listening performed by a component.
+
+To find out what this IP is on a Linux host you can use:
+
+```shell
+ip route show # Look for a line starting with "default via"
+```
+
+Kubernetes components do not accept custom network interface as an option,
+therefore a custom IP address must be passed as a flag to all components instances
+that need such a custom configuration.
+
+To configure the API server advertise address for control plane nodes created with both
+`init` and `join`, the flag `--apiserver-advertise-address` can be used.
+Preferably, this option can be set in the [kubeadm API](/docs/reference/config-api/kubeadm-config.v1beta3)
+as `InitConfiguration.localAPIEndpoint` and `JoinConfiguration.controlPlane.localAPIEndpoint`.
+
+For kubelets on all nodes, the `--node-ip` option can be passed in
+`.nodeRegistration.kubeletExtraArgs` inside a kubeadm configuration file
+(`InitConfiguration` or `JoinConfiguration`).
+
+{{< note >}}
+IP addresses become part of certificates SAN fields. Changing these IP addresses would require
+signing new certificates and restarting the affected components, so that the change in
+certificate files is reflected. See
+[Manual certificate renewal](/docs/tasks/administer-cluster/kubeadm/kubeadm-certs/#manual-certificate-renewal)
+for more details on this topic.
+{{</ note >}}
+
+{{< warning >}}
+The Kubernetes project recommends against this approach (configuring all component instances
+with custom IP addresses). Instead, the Kubernetes maintainers recommend to setup the host network,
+so that the default gateway IP is the one that Kubernetes components auto-detect and use.
+On Linux nodes, you can use commands such as `ip route` to configure networking; your operating
+system might also provide higher level network management tools. If your node's default gateway
+is a public IP address, you should configure packet filtering or other security measures that
+protect the nodes and your cluster.
+{{< /warning >}}
+
+{{< note >}}
+If the host does not have a default gateway, it is recommended to setup one. Otherwise,
+without passing a custom IP address to a Kubernetes component, the component
+will exit with an error. If two or more default gateways are present on the host,
+Kubernetes components will try to use the one that has a global unicast IP address
+(non-loopback, non-link local, non-point2point).
```

Comment:
We might want to document what happens when there are two of these (less common, but actually again this could apply to my laptop: I have wired ethernet and wireless and both interfaces have global unicast IPv6 addresses).

Reply:
the logic is quite complicated and a bit confusing - i.e. there are two methods.

### 1

if /proc/net/route does not exist it will enumerate interfaces using Go lang's net.Interfaces() which is OS portable. on Linux the enumeration is done with a set of POSIX syscalls, order should be the same as `ip route`. but i don't want to document such a thing as implementation can vary between kernels, CLI tools, etc. there is no guarantee of order determinism.

it would then do this:
```
// chooseIPFromHostInterfaces looks at all system interfaces, trying to find one that is up that
// has a global unicast address (non-loopback, non-link local, non-point2point), and returns the IP.
// addressFamilies determines whether it prefers IPv4 or IPv6
```
https://github.com/kubernetes/apimachinery/blob/12dc3f82eb47fc972b5be0b222f4422cb1b51612/pkg/util/net/interface.go#L309C1-L312

here we can say that the first IP on an interface that satisfies the above criteria is used.

### 2

if /proc/net/route exists both /proc/net/route  and  /proc/net/route_ipv6 are parsed with only default routes.
if the preferred IP family is v6, the v6 routes take precedence.

this is more deterministic. routes are appended in that same order.

then the IP is determined slightly differently, as there is a fallback to global IP on loopback:
```
// chooseHostInterfaceFromRoute cycles through each default route provided, looking for a
// global IP address from the interface for the route. If there are routes but no global
// address is obtained from the interfaces, it checks if the loopback interface has a global address.
// addressFamilies determines whether it prefers IPv4 or IPv6
```

https://github.com/kubernetes/apimachinery/blob/12dc3f82eb47fc972b5be0b222f4422cb1b51612/pkg/util/net/interface.go#L426-L430

@aojea can surely provide more context on this, but IMO what we can document is the following, trying to find a balance between somewhat descriptive and vague (to not enter rabbit holes):

```
If two or more default gateways are present on the host,
Kubernetes components will try to use the first one it encounters that has a suitable
global unicast IP address. While making this choice, the exact ordering of gateways
might vary between different operating systems and kernel versions.
```